src/conv/heuristic_model/tuning_heuristic.cpp

```diff
@@ -1,21 +1,145 @@
+
+        std::vector<float> decoder_input_vector(1, 0.0);
+        auto decoder_input_tensor = fdeep::tensor(fdeep::tensor_shape(1), decoder_input_vector);
+        std::vector<fdeep::tensor> decoder_input = {
+            decoder_input_tensor,
+            hidden_states[0],
+            hidden_states[1],
+            hidden_states[2],
+            hidden_states[3]}; // pass in SOS token and hidden states
+
+        for(int i = 0; i < metadata["num_tuning_params"].get<int>(); i++)
+        {
+            auto output        = decoder.predict({decoder_input});
+            auto output_vector = output[0].to_vector();
+            std::priority_queue<std::pair<float, int>> pq;
+            for(int j = 0; j < output_vector.size(); j++)
+                pq.push(std::make_pair(output_vector[j], j)); // sort by value at index
+
+            int output_token_index = -1;
+            while(!pq.empty())
+            {
+                int token = pq.top().second;
+                int value = metadata["decodings"]["tunings"][std::to_string(token)]
+                                .get<int>(); // convert index to tuning value
+                pq.pop();
+                if(value < 0)
+                    return false;
+                if(validator(i, value))
+                {
+                    output_token_index =
+                        token; // index with largest value that is valid = predicted index
+                    break;
+                }
+            }
+            decoder_input_tensor = fdeep::tensor(fdeep::tensor_shape(1),
+                                                 std::vector<float>(1, float(output_token_index)));
+            decoder_input        = {decoder_input_tensor,
+                             output[1],
+                             output[2],
+                             output[3],
+                             output[4]}; // index fed into decoder along with hidden states produced
+                                                // from previous step
+        }
+        return true;
+    }
+    size_t GetNumParams() const { return metadata["num_conv_params"].get<std::size_t>(); }
+    std::string arch;
+    std::string solver;
+    fdeep::model encoder;
+    fdeep::model decoder;
+    nlohmann::json metadata;
+};
+
+PerfTuningModel::PerfTuningModel() : pImpl(nullptr) {}
+PerfTuningModel::PerfTuningModel(const std::string& arch, const std::string& solver)
+    : pImpl{std::make_unique<impl>(arch, solver)}
+{
+}
+PerfTuningModel::~PerfTuningModel()                          = default;
+PerfTuningModel::PerfTuningModel(PerfTuningModel&&) noexcept = default;
+PerfTuningModel& PerfTuningModel::operator=(PerfTuningModel&&) noexcept = default;
+
+bool PerfTuningModel::ModelSetParams(std::function<bool(int, int)> validator,
+                                     const std::vector<float>& features) const
+{
+    return this->pImpl->ModelSetParams(validator, features);
```

Comment:
[Future] Pimpl idiom used. As I said before,

> As pimpl idiom introduces run-time overhead (indirection plus heap usage) and increases maintenance costs I will ask for more detailed justification (when the time comes, not now).

It's time. Why we need this?

[Notice] In spite of the above, the pimpl code has been silently added without notice. This is surprising.

Reply:
Because I am not aware of a cleaner way to remove the inlining of header-included functions. I am open to suggestions though